libraries/libopaemem/hash_map.c

```diff
@@ -0,0 +1,291 @@
+				key);
+
+	list = hm->buckets[key_hash];
+	if (!list) {
+		// Bucket was empty. Fill it.
+		hm->buckets[key_hash] = item;
+		return FPGA_OK;
+	}
+
+	// Bucket not empty. Do we have a key collision?
+	while (list) {
+		if (!hm->key_compare(key, list->key)) {
+			// Key collision.
+			opae_free(item);
+			if (hm->value_cleanup)
+				hm->value_cleanup(list->value, hm->cleanup_context);
+			list->value = value; // Replace value only.
+			return FPGA_OK;
+		}
+		list = list->next;
+	}
+
+	// No key collision. Add item to the list.
+	item->next = hm->buckets[key_hash];
+	hm->buckets[key_hash] = item;
+
+	return FPGA_OK;
+}
+
+fpga_result opae_hash_map_find(opae_hash_map *hm,
+			       void *key,
+			       void **value)
+{
+	uint32_t key_hash;
+	opae_hash_map_item *list;
+
+	if (!hm) {
+		ERR("NULL pointer");
+		return FPGA_INVALID_PARAM;
+	}
+
+	key_hash = hm->key_hash(hm->num_buckets,
+				hm->hash_seed,
+				key);
+
+	list = hm->buckets[key_hash];
+
+	while (list) {
+		if (!hm->key_compare(key, list->key)) {
+			if (value)
+				*value = list->value;
+			return FPGA_OK;
+		}
+		list = list->next;
+	}
+
+	return FPGA_NOT_FOUND;
+}
+
+fpga_result opae_hash_map_remove(opae_hash_map *hm,
+				 void *key)
+{
+	uint32_t key_hash;
+	opae_hash_map_item *prev = NULL;
+	opae_hash_map_item *list;
+
+	if (!hm) {
+		ERR("NULL pointer");
+		return FPGA_INVALID_PARAM;
+	}
+
+	key_hash = hm->key_hash(hm->num_buckets,
+				hm->hash_seed,
+				key);
+
+	list = hm->buckets[key_hash];
+	if (!list)
+		return FPGA_NOT_FOUND;
+
+	while (list) {
+		if (!hm->key_compare(key, list->key))
+			break;
+		prev = list;
+		list = list->next;
+	}
+
+	if (!list)
+		return FPGA_NOT_FOUND;
+
+	if (!prev) {
+		// Key found at head of list.
+		hm->buckets[key_hash] = list->next;
+	} else {
+		prev->next = list->next;
+	}
+
+	if (hm->key_cleanup)
+		hm->key_cleanup(list->key, hm->cleanup_context);
+	if (hm->value_cleanup)
+		hm->value_cleanup(list->value, hm->cleanup_context);
+	opae_free(list);
+
+	return FPGA_OK;
+}
+
+fpga_result opae_hash_map_destroy(opae_hash_map *hm)
+{
+	uint32_t i;
+
+	if (!hm) {
+		ERR("NULL pointer");
+		return FPGA_INVALID_PARAM;
+	}
+
+	for (i = 0 ; i < hm->num_buckets ; ++i) {
+		opae_hash_map_item *item;
+		item = hm->buckets[i];
+		while (item) {
+			opae_hash_map_item *trash;
+			trash = item;
+			item = item->next;
+			if (hm->key_cleanup)
+				hm->key_cleanup(trash->key, hm->cleanup_context);
+			if (hm->value_cleanup)
+				hm->value_cleanup(trash->value, hm->cleanup_context);
+			opae_free(trash);
+		}
+	}
+
+	opae_free(hm->buckets);
+	memset(hm, 0, sizeof(*hm));
+
+	return FPGA_OK;
+}
+
+bool opae_hash_map_is_empty(opae_hash_map *hm)
+{
+	uint32_t i;
+
+	for (i = 0 ; i < hm->num_buckets ; ++i) {
+		if (hm->buckets[i])
+			return false;
+	}
+
+	return true;
+}
+
+uint32_t opae_u64_key_hash(uint32_t num_buckets, uint32_t hash_seed, void *key)
+{
+	UNUSED_PARAM(hash_seed);
+	uint64_t remote_id = (uint64_t)key;
+	uint64_t hash = remote_id % 17659;
```

Comment:
What if someone wants > 17659 buckets? It would be better to just set num_buckets to prime. The way you have it now with 16384 buckets, you've set up a higher probability of using buckets between 0-1274 because of the way hash from line 276 will then map to num_buckets in 277.

Reply:
Revised the hash function logic and creating num_buckets with a prime.